### PR TITLE
Add support for gather/scatter of dataclass

### PR DESCRIPTION
**Pull request by rtabbara**
Motivated by #303 , add support for `gather` and `scatter` when using `dataclass` types

### CODE REVIEW

**Comment by merlinND** [comment]
Very cool!
Could you please check if this also works with `__future__`-style annotations?
In the unit tests, that might require a separate `test_*.py` file since `from __future__ import annotations` affects the whole file I think.

**Comment by rtabbara** [comment]
Hey @merlinND 

I've tested it out locally and it works but with caveats. There's a more general issue that was uncovered when using `from __future__ import annotations` to lazily resolve the type. Internally, we're using `typing.get_type_hints`, where by default the "appropriate namespace dictionaries are inferred from obj" when selecting the global and local namespaces used for type lookup. The issue is if for instance you have some inner defined type


```python
def my_func(t):

    @dataclass
    class MyDataclass:
        a : t
```

where `t` itself is not part of the global namespace, then `typing.get_type_hints(MyDataclass)` won't be able to resolve the type `t` because it's not part of its local namespace either.

In short, at least for the purposes of Pytest, I'm not sure we can really modify the tests to accommodate for this use case because the tests are exactly setup like this where the types are passed in arguments.

In practice, I'm not too sure how big a problem this is because as I understand, the main motivation was when using Mitsuba where we can define dataclasses containing fields with alias types (`mi.Float`) before the variant has been set. And if you're adding Mitsuba to the global namespace then it should be fine.

Anyway, I hope that's not too much of an annoyance but please let me know otherwise.

**Comment by merlinND** [comment]
Hi @rtabbara,

Thanks a lot for checking and the careful explanations.

I understand why `: t` wouldn't work with future-style annotations, it makes sense and I don't think it's a problem.
Actually if I try something like the snippet you shared, Pylance warns me that I'm not supposed to use a variable type in annotations :stuck_out_tongue: 

If we really wanted to add a unit test for future-style annotations support, it could be possible with a separate test file I think, something like:

```py
from __future__ import annotations

@pytest.test_arrays('TODO')
def test01_dataclass_with_postponed_types(t):
    if dr.backend_v(t) == dr.JitBackend.CUDA:
        @dataclass
        class MyDataclass:
            a : dr.cuda.Float16
    elif dr.backend_v(t) == dr.JitBackend.LLVM:
        @dataclass
        class MyDataclass:
            a : dr.llvm.Float16

    ... = dr.full(MyDataclass, 10)
    ... = dr.gather(MyDataclass, ...)
    dr.enable_grad(MyDataclass)
    ...
```

**Review by njroussel — APPROVED** [review]
Looks good to me!
